[docs] Document tips for debugging assembly & object code.

varungandhi-apple opened the pull request:
Based on some recent adventures and conversations with Mike. 🙂

varungandhi-apple committed:
[docs] Document tips for debugging assembly & object code.

mikeash commented:
Looks good, thanks for taking the time to write this stuff down.

kastiglione commented on docs/DebuggingTheCompiler.md:
```diff
@@ -553,6 +553,95 @@ the pipeline. Here is a quick summary of the various options:
+nm -n -m tmp.o | swift demangle > tmp.txt
+
+# Inspect disassembly of an existing dylib (AT&T syntax is the default)
+objdump -d --x86-asm-syntax=intel /path/to/libcake.dylib \
+  | swift demangle > libcake.S
```
I vote for recommending `-macho` or simply `-m`, the output is much better for macho/Darwin anyway. Specifically it includes comments like:

```
1000032b4:     e8 35 09 00 00  call    0x100003bee ## symbol stub for: _$s12_Concurrency9MainActorCMa
```

kastiglione replied:
One of us (or someone) should wire up libswiftDemangle into `llvm-nm` because it has a `--demangle` flag, but that doesn't seem to work on swift symbols.

varungandhi-apple replied:
Nice, I will add the `-macho` recommendation. 👍🏼

Hmm, when I do `man nm`, it doesn't have a `--demangle` flag. Maybe it's just undocumented? It certainly seems to work on C++ code. 🤔

varungandhi-apple replied:
For the libswiftDemangle idea, I think it would be nice-to-have, but I'm not sure if the benefit is worth the implementation effort (maybe I'm overestimating the complexity of implementation though). I think the current workflow is good enough for my use case.

kastiglione replied:
the `--demangle` flag is for `objdump`

varungandhi-apple commented on docs/DebuggingTheCompiler.md:
````diff
@@ -553,6 +553,95 @@ the pipeline. Here is a quick summary of the various options:
+  | swift demangle > libcake.S
+```
+
+### Working with multiple files
+
+Some bugs only manifest in WMO, and may involve complicated Xcode projects.
+Moreover, Xcode may be passing arguments via `-filelist`
+and expecting outputs via `-output-filelist`, and those file lists
+may be in temporary directories.
+
+If you want to inspect assembly or object code for individual files when
+compiling under WMO, you can mimic this by doing the following:
````
Right, by default WMO will use a single object file. Xcode is ensuring that N object files are created by passing in a `-output-file-map /path/toProjectName-OutputFileMap.json` which contains paths for the various files such as diagnostic files, dependency files and object files. Since the JSON format is a bit more work to create, I thought it would be better to stick to the simpler text based format. However, `-output-filelist` is a frontend flag, which is why I suggest invoking the frontend directly here.

kastiglione replied:
In the past I've created a tool to generate an output file map for some given source files, and the output paths used convention. Being a tool it also could create intermediate directories. I think it could be reasonable to include such a tool, or even put some simple-case logic into the driver. But at this point in time, I guess the frontend and these files are the way to go.

varungandhi-apple commented on docs/DebuggingTheCompiler.md:
```diff
@@ -553,6 +553,95 @@ the pipeline. Here is a quick summary of the various options:
+sed -e 's|.*/|Output/|;s|\.swift|.S|' input-files.txt > output-files.txt
+
+# Paste rest of invocation from Xcode's 'CompileSwiftSources' phase
+# in the build log and remove the existing -filelist and -output-filelist
```
This is a good point 👍🏼, I checked that the compiler uses the last argument, so I can simplify this.

varungandhi-apple commented on docs/DebuggingTheCompiler.md:
```diff
@@ -553,6 +553,95 @@ the pipeline. Here is a quick summary of the various options:
+# Paste rest of invocation from Xcode's 'CompileSwiftSources' phase
+# in the build log and remove the existing -filelist and -output-filelist
+# arguments.
+xcrun swift-frontend <... other args ...> \
```
Do you mean suggesting something like the following:

```
// compiler-args.txt
-sdk /path/to/sdk <... args ...>

echo "$(cat compiler-args.txt) -filelist input-files.txt -output-filelist output-files.txt -O -whole-module-optimization -emit-assembly" | xargs /path/to/swift-frontend
```

Or something else?

It seems like a reasonable alternative. Generally my workflow here is to save the invocation itself to a shell script, and edit it there. That also makes it easier to add post-processing steps, such as demangling all the outputs. Or demangling one file and grepping the output.

kastiglione replied:
Save the files to `args.txt`, and then run `swift-frontend @args.txt -extra xxx -flags yyy`.

kastiglione replied:
For some small context, see https://llvm.org/docs/CommandLine.html#response-files

varungandhi-apple replied:
Whaaaaat, I thought the `@` indicated a filelist, I didn't realize you could substitute arbitrary arguments using it. 🤯

varungandhi-apple commented:
Leaving this open for a day, in case anyone has ideas/suggestions later. Will smoke test and merge tomorrow.

kastiglione reviewed (approved):
I wonder if we should also maybe include a table of sometimes useful tools. They wouldn't even need to be documented much, maybe just their name, and a one liner about it. It can also be good as an overview. Including what's already mentioned in this PR, here's a list I can think of:

* `install_name_tool`
* `llvm-bcanalyzer`
* `c++filt`
* `dyldinfo`
* `nm`
* `objdump` (`otool`?)
* `swift demangle`
* `dwarfdump`
* `lipo`
* `size`
* `strings`
* `libtool`

varungandhi-apple commented:
I really like this idea. Instead of just mentioning names, having a couple of sentences or two would be useful. Bit occupied at the moment, I will get to this next week. 👍🏼

MaxDesiatov commented:
@swift-ci please smoke test

varungandhi-apple commented:
FYI @MaxDesiatov , this isn't ready to be merged, I have some changes locally, I need to finish those and re-push.

MaxDesiatov commented:
I noticed that only after triggering CI, sorry.

varungandhi-apple committed:
[docs] Mention useful debugging tools with short blurbs about usage.

Based on Dave Lee's suggestion in:
https://github.com/apple/swift/pull/35639#pullrequestreview-579556113

varungandhi-apple commented:
@kastiglione I've added short descriptions for the tools you suggested. 😄

kastiglione commented on docs/DebuggingTheCompiler.md:
```diff
@@ -553,6 +553,149 @@ the pipeline. Here is a quick summary of the various options:
+- Linking:
+  - `libtool`: A tool to create static and dynamic libraries. Generally, it's
+    easier to instead ask `swiftc` to link files, but potentially handy as
+    a higher-level alternative to `ld` and `lipo`.
```
should `ar` be mentioned? ex:

> alternative to `ld`, `ar`, and `lipo`?

varungandhi-apple replied:
Added reference to `ar`.

kastiglione commented on docs/DebuggingTheCompiler.md:
```diff
@@ -553,6 +553,149 @@ the pipeline. Here is a quick summary of the various options:
+    easier to instead ask `swiftc` to link files, but potentially handy as
+    a higher-level alternative to `ld` and `lipo`.
+
+- Debuginfo:
```
Is this a one-worder? I usually spell it "debug info"

varungandhi-apple replied:
I wasn't sure. 😅 Seems like debug info is much more common.

kastiglione commented on docs/DebuggingTheCompiler.md:
```diff
@@ -553,6 +553,149 @@ the pipeline. Here is a quick summary of the various options:
+      loaded by a running application, you can obtain the UUID using
+      `image list` in LLDB.
+- `objdump`: Dump object files.
+   Some examples of using `objdump` are documented in the previous subsection.
```
For me, `objdump` is a commonly used tool and a bit of a Swiss army knife. Is it worth selling it a bit more, for example mentioning that it has a lot of features and use cases compared to the other tools in the list.

varungandhi-apple replied:
Added a sentence about objdump's functionality.

kastiglione reviewed (commented):
a few comments, but looks really good. Great additions

varungandhi-apple commented:
Michael G. suggested I add descriptions of which tools work on different platforms vs which tools work only on 1 platform; unfortunately, I don't have time for diving more into that right now, so I've left a TODO for it with a description of how it could look like.

varungandhi-apple commented:
@swift-ci smoke test

varungandhi-apple commented:
Going to merge this; in case there are any other suggestions, I can fix them in a follow-up PR.

kastiglione commented:
this is great
